src/string_array.c

```diff
@@ -133,6 +134,78 @@ rcutils_string_array_cmp(
   return RCUTILS_RET_OK;
 }
 
+rcutils_ret_t
+rcutils_string_array_resize(
+  rcutils_string_array_t * string_array,
+  size_t new_size)
+{
+  RCUTILS_CHECK_FOR_NULL_WITH_MSG(
+    string_array, "string_array is null", return RCUTILS_RET_INVALID_ARGUMENT);
+
+  if (string_array->size == new_size) {
+    return RCUTILS_RET_OK;
+  }
+
+  if (0 == new_size) {
+    return rcutils_string_array_fini(string_array);
+  }
+
+  // Reclaim entries being removed
+  rcutils_allocator_t * allocator = &string_array->allocator;
+  for (size_t i = new_size; i < string_array->size; ++i) {
+    allocator->deallocate(string_array->data[i], allocator->state);
+    string_array->data[i] = NULL;
+  }
+
+  char ** new_data = allocator->reallocate(
+    string_array->data, new_size * sizeof(char *), allocator->state);
+  if (NULL == new_data) {
+    RCUTILS_SET_ERROR_MSG("failed to allocator string array");
+    return RCUTILS_RET_BAD_ALLOC;
+  }
+  string_array->data = new_data;
+
+  // Zero-initialize new entries
+  for (size_t i = string_array->size; i < new_size; ++i) {
+    string_array->data[i] = NULL;
+  }
+
+  string_array->size = new_size;
+
+  return RCUTILS_RET_OK;
+}
+
+static int
+rcutils_string_array_sort_compare(const void * lhs, const void * rhs)
+{
+  const char * left = *(const char **)lhs;
+  const char * right = *(const char **)rhs;
+  if (NULL == left) {
+    return NULL == right ? 0 : 1;
+  } else if (NULL == right) {
+    return -1;
+  }
+  return strcmp(left, right);
+}
+
+rcutils_ret_t
+rcutils_string_array_sort(rcutils_string_array_t * string_array)
+{
+  RCUTILS_CHECK_FOR_NULL_WITH_MSG(
+    string_array, "string_array is null", return RCUTILS_RET_INVALID_ARGUMENT);
+
+  if (1 >= string_array->size) {
+    return RCUTILS_RET_OK;
+  }
+
+  RCUTILS_CHECK_FOR_NULL_WITH_MSG(
+    string_array->data, "string_array->data is null", return RCUTILS_RET_INVALID_ARGUMENT);
```

Comment:
```suggestion
  RCUTILS_CHECK_FOR_NULL_WITH_MSG(
    string_array->data, "string_array->data is null", return RCUTILS_RET_INVALID_ARGUMENT);

    if (1 >= string_array->size) {
    return RCUTILS_RET_OK;
  }
```

if data is NULL, sorting cannot be done. so i believe arguments should be checked at entry point.

Reply:
I would argue that a `string_array` of size 0 is already sorted, and we don't need to touch the data in a string array of size 1 either, so it seems like more optimized code to keep the `NULL` check where it is.

Also, it seems that a `string_array` of size 0 can have a `NULL` `data` value in some cases, so it would be best not to disclude that case.

Reply:
> I would argue that a string_array of size 0 is already sorted

i guess that is okay, thanks